### PR TITLE
Generalize accessor storage to preserve the original accessor list.

### PR DESCRIPTION
**Pull request by rjmccall**
Only not NFC because it's detectable by source tools.

### CODE REVIEW

**Comment by rjmccall** [comment]
@swift-ci Please test.

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -4113,66 +4115,65 @@ class AbstractStorageDecl : public ValueDecl {
...
+
+    /// The storage capacity of this record for accessors.  Always includes
+    /// enough space for adding opaque accessors to the record, which are the
+    /// only accessors that should ever be added retroactively; hence this
+    /// field is only here for the purposes of safety checks.
+    AccessorIndex AccessorsCapacity;
+
+    /// Either 0, meaning there is no registered accessor of the given kind,
+    /// or the index+1 of the accessor in the accessors array.
+    AccessorIndex AccessorIndices[NumAccessorKinds];
```
Comment (jrose-apple) [inline include/swift/AST/Decl.h]: Can you use llvm::TrailingObjects instead of a supposed fixed-sized array?

Comment (rjmccall) [reply include/swift/AST/Decl.h]: That is a fixed-size array; it's indexed by AccessorKinds.  There *is* a trailing-objects use in that type, though.

Comment (jrose-apple) [reply include/swift/AST/Decl.h]: Oops, my brain misread NumAccessorKinds as MaxNumAccessors. Sorry for the noise!

Comment (rjmccall) [reply include/swift/AST/Decl.h]: I've updated to use TrailingObjects for the part that's actually a trailing array.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
Test failure is a consequence of other changes.  Rebasing and resubmitting.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

File: test/SourceKit/InterfaceGen/gen_clang_module.swift.response [hunk 2]
```diff
@@ -4367,13 +4367,14 @@ public class FooOverlayClassDerived : Foo.FooOverlayClassBase {
   {
     key.kind: source.lang.swift.decl.var.global,
     key.accessibility: source.lang.swift.accessibility.public,
-    key.setter_accessibility: source.lang.swift.accessibility.public,
```
[the remark this answers is not in the document]
Comment (benlangmuir) [reply test/SourceKit/InterfaceGen/gen_clang_module.swift.response]: LGTM

**Pull request merged**